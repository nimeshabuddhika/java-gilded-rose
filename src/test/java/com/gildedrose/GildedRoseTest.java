package com.gildedrose;

import org.approvaltests.combinations.CombinationApprovals;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class GildedRoseTest {

    /**
     * 64 number of test cases
     */
    @Test
    void allCombinations() {
        String[] name = {"foo", "Aged Brie", "Backstage passes to a TAFKAL80ETC concert", "Sulfuras, Hand of Ragnaros"};
        Integer[] sellIn = {-1, 0, 5, 12};
        Integer[] quality = {0, 1, 49, 55};

        CombinationApprovals.verifyAllCombinations(this::updateQuality, name, sellIn, quality);
    }

    private String updateQuality(String name, int sellIn, int quality) {
        GildedRose app = new GildedRose(new Item[]{new Item(name, sellIn, quality)});
        app.updateQuality();
        return app.items[0].toString();
    }

    @Test
    void conjuredTwiceSpeed() {
        GildedRose app = new GildedRose(new Item[]{new Item("Conjured Mana Cake", 3, 10)});

        app.updateQuality();

        assertEquals(app.items[0].toString(), new Item("Conjured Mana Cake", 2, 8).toString());
    }

    @Test
    void conjuredTwiceSpeedAndOutOfStock() {
        GildedRose app = new GildedRose(new Item[]{new Item("Conjured Mana Cake", 0, 10)});

        app.updateQuality();

        assertEquals(app.items[0].toString(), new Item("Conjured Mana Cake", -1, 6).toString());
    }

    @Test
    void conjuredTwiceSpeedAndZeroQuality() {
        GildedRose app = new GildedRose(new Item[]{new Item("Conjured Mana Cake", 0, 0)});

        app.updateQuality();

        assertEquals(app.items[0].toString(), new Item("Conjured Mana Cake", -1, 0).toString());
    }

}
