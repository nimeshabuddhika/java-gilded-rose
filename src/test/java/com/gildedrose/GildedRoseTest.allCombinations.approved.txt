[foo, -1, 0] => foo, -2, 0 
[foo, -1, 1] => foo, -2, 0 
[foo, -1, 49] => foo, -2, 47 
[foo, -1, 55] => foo, -2, 53 
[foo, 0, 0] => foo, -1, 0 
[foo, 0, 1] => foo, -1, 0 
[foo, 0, 49] => foo, -1, 47 
[foo, 0, 55] => foo, -1, 53 
[foo, 5, 0] => foo, 4, 0 
[foo, 5, 1] => foo, 4, 0 
[foo, 5, 49] => foo, 4, 48 
[foo, 5, 55] => foo, 4, 54 
[foo, 12, 0] => foo, 11, 0 
[foo, 12, 1] => foo, 11, 0 
[foo, 12, 49] => foo, 11, 48 
[foo, 12, 55] => foo, 11, 54 
[Aged Brie, -1, 0] => Aged Brie, -2, 2 
[Aged Brie, -1, 1] => Aged Brie, -2, 3 
[Aged Brie, -1, 49] => Aged Brie, -2, 50 
[Aged Brie, -1, 55] => Aged Brie, -2, 55 
[Aged Brie, 0, 0] => Aged Brie, -1, 2 
[Aged Brie, 0, 1] => Aged Brie, -1, 3 
[Aged Brie, 0, 49] => Aged Brie, -1, 50 
[Aged Brie, 0, 55] => Aged Brie, -1, 55 
[Aged Brie, 5, 0] => Aged Brie, 4, 1 
[Aged Brie, 5, 1] => Aged Brie, 4, 2 
[Aged Brie, 5, 49] => Aged Brie, 4, 50 
[Aged Brie, 5, 55] => Aged Brie, 4, 55 
[Aged Brie, 12, 0] => Aged Brie, 11, 1 
[Aged Brie, 12, 1] => Aged Brie, 11, 2 
[Aged Brie, 12, 49] => Aged Brie, 11, 50 
[Aged Brie, 12, 55] => Aged Brie, 11, 55 
[Backstage passes to a TAFKAL80ETC concert, -1, 0] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 1] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 49] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 55] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 0] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 1] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 49] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 55] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 5, 0] => Backstage passes to a TAFKAL80ETC concert, 4, 3 
[Backstage passes to a TAFKAL80ETC concert, 5, 1] => Backstage passes to a TAFKAL80ETC concert, 4, 4 
[Backstage passes to a TAFKAL80ETC concert, 5, 49] => Backstage passes to a TAFKAL80ETC concert, 4, 50 
[Backstage passes to a TAFKAL80ETC concert, 5, 55] => Backstage passes to a TAFKAL80ETC concert, 4, 55 
[Backstage passes to a TAFKAL80ETC concert, 12, 0] => Backstage passes to a TAFKAL80ETC concert, 11, 1 
[Backstage passes to a TAFKAL80ETC concert, 12, 1] => Backstage passes to a TAFKAL80ETC concert, 11, 2 
[Backstage passes to a TAFKAL80ETC concert, 12, 49] => Backstage passes to a TAFKAL80ETC concert, 11, 50 
[Backstage passes to a TAFKAL80ETC concert, 12, 55] => Backstage passes to a TAFKAL80ETC concert, 11, 55 
[Sulfuras, Hand of Ragnaros, -1, 0] => Sulfuras, Hand of Ragnaros, -1, 0 
[Sulfuras, Hand of Ragnaros, -1, 1] => Sulfuras, Hand of Ragnaros, -1, 1 
[Sulfuras, Hand of Ragnaros, -1, 49] => Sulfuras, Hand of Ragnaros, -1, 49 
[Sulfuras, Hand of Ragnaros, -1, 55] => Sulfuras, Hand of Ragnaros, -1, 55 
[Sulfuras, Hand of Ragnaros, 0, 0] => Sulfuras, Hand of Ragnaros, 0, 0 
[Sulfuras, Hand of Ragnaros, 0, 1] => Sulfuras, Hand of Ragnaros, 0, 1 
[Sulfuras, Hand of Ragnaros, 0, 49] => Sulfuras, Hand of Ragnaros, 0, 49 
[Sulfuras, Hand of Ragnaros, 0, 55] => Sulfuras, Hand of Ragnaros, 0, 55 
[Sulfuras, Hand of Ragnaros, 5, 0] => Sulfuras, Hand of Ragnaros, 5, 0 
[Sulfuras, Hand of Ragnaros, 5, 1] => Sulfuras, Hand of Ragnaros, 5, 1 
[Sulfuras, Hand of Ragnaros, 5, 49] => Sulfuras, Hand of Ragnaros, 5, 49 
[Sulfuras, Hand of Ragnaros, 5, 55] => Sulfuras, Hand of Ragnaros, 5, 55 
[Sulfuras, Hand of Ragnaros, 12, 0] => Sulfuras, Hand of Ragnaros, 12, 0 
[Sulfuras, Hand of Ragnaros, 12, 1] => Sulfuras, Hand of Ragnaros, 12, 1 
[Sulfuras, Hand of Ragnaros, 12, 49] => Sulfuras, Hand of Ragnaros, 12, 49 
[Sulfuras, Hand of Ragnaros, 12, 55] => Sulfuras, Hand of Ragnaros, 12, 55 
